src/vset.c

```diff
@@ -0,0 +1,2084 @@
+ *         If the resulting vector size drops to 1, it is converted to a SINGLE bucket.
+ *         If the vector becomes empty, it is removed entirely (set to NONE).
+ *
+ *     - VSET_BUCKET_RAX:
+ *         The function decodes the appropriate bucket key (based on the expiration
+ *         time), looks up the RAX node, and dispatches removal to the sub-bucket.
+ *         If a sub-bucket becomes empty or has only one entry left, its bucket
+ *         type may be downgraded (e.g., to SINGLE or removed).
+ *
+ *  3. If the removal results in a structural change (e.g., shrinking a bucket),
+ *     the bucket type may be changed, and the root pointer is updated accordingly.
+ *
+ *  4. If the entry is not found in the expected bucket, no action is taken.
+ *
+ * Notes:
+ *  - Buckets self-adjust during removal for memory efficiency.
+ *  - The vector bucket keeps entries sorted for fast search/removal.
+ *  - RAX-based sets support a large number of buckets and scale well
+ *    with many time windows.
+ *  - Entries are assumed to have pointer identity (odd-valued pointers).
+ *  - Correct expiration timestamp must be provided for accurate removal.
+ *
+ * Return value:
+ *     Returns true if the entry was found and removed successfully.
+ *     Returns false if the entry was not found.
+ *
+ * Example usage:
+ *     vsetRemoveEntry(myset, extract_expiry, my_object);
+ *
+ *     // my_object is removed from the appropriate bucket in myset BUT is not freed. */
+bool vsetRemoveEntry(vset *set, vsetGetExpiryFunc getExpiry, void *entry) {
+    return vsetRemoveEntryWithExpiry(set, getExpiry, entry, getExpiry(entry));
+}
+
+static inline vsetBucket *vsetBucketUpdateEntry_SINGLE(vsetBucket *bucket, vsetGetExpiryFunc getExpiry, void *old_entry, void *new_entry, long long old_expiry, long long new_expiry) {
+    UNUSED(getExpiry);
+    UNUSED(old_expiry);
+    UNUSED(new_expiry);
+
+    if (vsetBucketSingle(bucket) == old_entry) {
+        return vsetBucketFromSingle(new_entry);
+    }
+    return vsetBucketFromNone();
+}
+
+static inline vsetBucket *vsetBucketUpdateEntry_VECTOR(vsetBucket *bucket, vsetGetExpiryFunc getExpiry, void *old_entry, void *new_entry, long long old_expiry, long long new_expiry) {
+    UNUSED(getExpiry);
+    UNUSED(old_expiry);
+    UNUSED(new_expiry);
+
+    pVector *pv = vsetBucketVector(bucket);
+    uint32_t idx = pvFind(pv, old_entry);
+    /* in case we did not locate the entry, just return NONE bucket */
+    if (idx == pvLen(pv))
+        return vsetBucketFromNone();
+    pvSet(pv, idx, new_entry);
+    return bucket;
+}
+
+static inline vsetBucket *vsetBucketUpdateEntry_HASHTABLE(vsetBucket *bucket, vsetGetExpiryFunc getExpiry, void *old_entry, void *new_entry, long long old_expiry, long long new_expiry) {
+    UNUSED(getExpiry);
+    UNUSED(old_expiry);
+    UNUSED(new_expiry);
+
+    hashtable *ht = vsetBucketHashtable(bucket);
+    void **ref = hashtableFindRef(ht, old_entry);
```

Comment:
Am I right that this corrupts the hashtable if new_entry has a different address than old_entry (and it should in most cases)? Since the hash is based on the pointer (`hash_pointer(ptr)`), the new entry likely hashes to a different bucket. Overwriting in-place (*ref = new_entry) breaks the bucket invariant and may lead to lookup failures and corruption. Shouldn’t we delete the old entry and re-insert the new one instead?
eg:
```
    if (hashtableDelete(ht, old_entry)) {
        hashtableAdd(ht, new_entry);
    }else {
        return vsetBucketFromNone();
    }
```

Reply:
super great point. you are right! the old entry should be removed

Reply:
fixed @xbasel thank you for locating this!!!